services/motion/builtin/state/state.go

```diff
@@ -52,16 +67,529 @@ type PlannerExecutor interface {
 	Cancel()
 }
 
+type componentState struct {
+	executionIDHistory []motion.ExecutionID
+	executionsByID     map[motion.ExecutionID]stateExecution
+}
+
+type newPlanMsg struct {
+	plan       motion.Plan
+	planStatus motion.PlanStatus
+}
+
+type stateUpdateMsg struct {
+	componentName resource.Name
+	executionID   motion.ExecutionID
+	planID        motion.PlanID
+	planStatus    motion.PlanStatus
+}
+
+// a stateExecution is the struct held in the state that
+// holds the history of plans & plan status updates an
+// execution has exprienced & the waitGroup & cancelFunc
+// required to shut down an execution's goroutine.
+type stateExecution struct {
+	id            motion.ExecutionID
+	componentName resource.Name
+	waitGroup     *sync.WaitGroup
+	cancelFunc    context.CancelFunc
+	history       []motion.PlanWithStatus
+}
+
+func (e *stateExecution) stop() {
+	e.cancelFunc()
+	e.waitGroup.Wait()
+}
+
+func (cs componentState) lastExecution() stateExecution {
+	return cs.executionsByID[cs.lastExecutionID()]
+}
+
+func (cs componentState) lastExecutionID() motion.ExecutionID {
+	return cs.executionIDHistory[0]
+}
+
+// execution represents the state of a motion planning execution.
+// it only ever exists in state.StartExecution function & the go routine created.
+type execution[R any] struct {
+	id                      motion.ExecutionID
+	state                   *State
+	waitGroup               *sync.WaitGroup
+	cancelCtx               context.Context
+	cancelFunc              context.CancelFunc
+	logger                  logging.Logger
+	componentName           resource.Name
+	req                     R
+	planExecutorConstructor PlanExecutorConstructor[R]
+}
+
+type planWithExecutor struct {
+	plan         motion.Plan
+	planExecutor PlannerExecutor
```

Comment:
can we change `PlannerExecutor` to `PlanExecutor`?

Reply:
done